File: services/ml-service/app/services/image_embeddings.py
"""
Image embedding generation using CLIP
"""
import base64
import io
import torch
from typing import List
from PIL import Image
from app.core.models import get_image_model
from app.core.config import settings


def decode_base64_image(base64_str: str) -> Image.Image:
    """
    Decode base64 string to PIL Image
    
    Args:
        base64_str: Base64 encoded image string
        
    Returns:
        PIL Image
    """
    # Remove data URL prefix if present
    if ',' in base64_str:
        base64_str = base64_str.split(',', 1)[1]
    
    # Decode base64
    image_bytes = base64.b64decode(base64_str)
    
    # Open as PIL Image
    image = Image.open(io.BytesIO(image_bytes))
    
    # Convert to RGB if needed
    if image.mode != 'RGB':
        image = image.convert('RGB')
    
    return image


async def generate_image_embeddings(images: List[str]) -> List[List[float]]:
    """
    Generate embeddings for a list of images
    
    Args:
        images: List of base64 encoded image strings
        
    Returns:
        List of embedding vectors (normalized)
    """
    if not images:
        return []
    
    model, processor = get_image_model()
    device = torch.device(settings.DEVICE)
    
    # Decode base64 images to PIL
    pil_images = []
    for img_str in images:
        try:
            img = decode_base64_image(img_str)
            pil_images.append(img)
        except Exception as e:
            raise ValueError(f"Failed to decode image: {str(e)}")
    
    # Process images
    inputs = processor(
        images=pil_images,
        return_tensors="pt",
        padding=True
    ).to(device)
    
    # Generate embeddings
    with torch.no_grad():
        image_features = model.get_image_features(**inputs)
    
    # Normalize embeddings
    image_features = torch.nn.functional.normalize(image_features, p=2, dim=1)
    
    # Convert to list
    embeddings_list = image_features.cpu().numpy().tolist()
    
    return embeddings_list


async def generate_image_embedding(image: str) -> List[float]:
    """
    Generate embedding for a single image
    
    Args:
        image: Base64 encoded image string
        
    Returns:
        Embedding vector (normalized)
    """
    embeddings = await generate_image_embeddings([image])
    return embeddings[0]

File: services/model-worker/src/services/image-embeddings.ts
import { getImageModel } from '../models/loader.js';
import { env } from '../config/env.js';
import sharp from 'sharp';

/**
 * Decode base64 image to buffer
 */
function decodeBase64Image(base64: string): Buffer {
  // Remove data URL prefix if present
  const base64Data = base64.replace(/^data:image\/\w+;base64,/, '');
  return Buffer.from(base64Data, 'base64');
}

/**
 * Preprocess image for CLIP model
 * - Resize to 224x224
 * - Convert to RGB
 */
async function preprocessImage(buffer: Buffer): Promise<Buffer> {
  try {
    return await sharp(buffer)
      .resize(224, 224, {
        fit: 'cover',
        position: 'center',
      })
      .removeAlpha()
      .toColorspace('srgb')
      .raw()
      .toBuffer();
  } catch (error) {
    throw new Error(`Image preprocessing failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

/**
 * Generate image embeddings using CLIP model
 * 
 * @param images - Array of base64 encoded images
 * @returns Array of embedding vectors
 */
export async function generateImageEmbeddings(images: string[]): Promise<number[][]> {
  if (images.length === 0) {
    return [];
  }

  // Validate batch size
  if (images.length > env.MAX_BATCH_SIZE) {
    throw new Error(`Batch size ${images.length} exceeds maximum ${env.MAX_BATCH_SIZE}`);
  }

  const model = getImageModel();

  try {
    // Decode and preprocess all images
    const imageBuffers = await Promise.all(
      images.map(async (base64) => {
        const buffer = decodeBase64Image(base64);
        return await preprocessImage(buffer);
      })
    );

    // Generate embeddings
    // Note: transformers.js expects image URLs or buffers
    // For buffers, we need to convert to a format the model accepts
    const embeddings: number[][] = [];

    for (const buffer of imageBuffers) {
      // Convert buffer to data URL for the model
      const base64 = buffer.toString('base64');
      const dataUrl = `data:image/raw;base64,${base64}`;
      
      const output = await model(dataUrl);
      const embedding = Array.from(output.data as Float32Array);
      embeddings.push(embedding);
    }

    return embeddings;
  } catch (error) {
    console.error('Image embedding generation failed:', error);
    throw new Error(`Failed to generate image embeddings: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

/**
 * Generate single image embedding
 * 
 * @param image - Base64 encoded image
 * @returns Embedding vector
 */
export async function generateImageEmbedding(image: string): Promise<number[]> {
  const embeddings = await generateImageEmbeddings([image]);
  return embeddings[0]!;
}

